proposals/Scoped-Custom-Element-Registries.md

```diff
@@ -0,0 +1,227 @@
+  constructor() {
+    this.attachShadow({mode: 'open', registry});
+  }
+}
+```
+
+### Scoped element creation APIs
+
+Element creation APIs, like `createElement()` and `innerHTML` can be grouped into global API (those on Document or Window) and scoped APIs (those on HTMLElement and ShadowRoot). The scoped APIs have an associated Node that can be used to look up a CustomElementRegistry and thus a scoped definition.
+
+In order to support scoped registries we add new scoped APIs, that were previously only available on `Document`, to `ShadowRoot`:
+
+* `createElement()`
+* `createElementNS()`
+* `importNode()`
+
+These APIs work the same as their Document equivalents, but use scoped registries instead of the global registry.
+
+### Registry Inheritance
+
+In a shadow root with a scoped registry, all element creation APIs use that shadow root's registry, and not the document's, to look up definitions. To inherit definitions from the global, or another scoped registry, a parent registry can be passed in at construction time:
+
+```js
+const registry = new CustomElementRegistry({
+    parent: window.customElements,
+});
+```
+
+This inheritance is live. New registrations added to the parent registry are available to inheriting registries. This is useful in the case where an element is already written to use the global registry, but needs to register a private helper element, or override only a single element in conflict with the global. It may also be useful in plug-in architectures where the host program provides a number of elements to plugins.
+
+For non-live inheritance, we can add a method to CustomElementRegistry that returns all of its registrations:
+
+```js
+const registry = new CustomElementRegistry({
+    definitions: {
+      ...window.customElements.getDefinitions(),
+      'local-element': LocalElement,
+    }
+});
+```
+
+### Finding a custom element definition
+
+Because there is no longer a single global custom element registry, when creating elements, the steps to look up a custom element definition need to be updated to be able to find the correct registry.
+
+That process needs to take a context node that is used to look up the definition. The registry is found by getting the context node's root. If the root has a CustomElementRegistry, use that registry to look up the definition, otherwise use the global objects CustomElementRegistry object.
+
+The context node is the node that hosts the element creation API that was invoked, such as `ShadowRoot.prototype.createElement()`, or `HTMLElement.prototype.innerHTML`. For `ShadowRoot.prototype.createElement()`, the context node and root are the same.
+
+#### Note on looking up registries
+
+One consequence of looking up a registry from the root at element creation time is that different registries could be used over time for some APIs like HTMLElement.prototype.innerHTML, if the context node moves between shadow roots. This should be exceedingly rare though.
```

Comment:
this sentence was a little confusing at first, maybe need some rewording!

Reply:
I'll try to clean this up.